sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -69,6 +73,33 @@ object DefaultOptimizer extends Optimizer {
 }
 
 /**
+ * Pushes down Limit for reducing the amount of returned data.
+ *
+ * 1. Adding Extra Limit beneath the operations, including Union All.
+ * 2. Project is pushed through Limit in the rule ColumnPruning
+ *
+ * Any operator that a Limit can be pushed passed should override the maxRows function.
+ *
+ * Note: This rule has to be done when the logical plan is stable;
+ *       Otherwise, it could impact the other rules.
```

Comment:
For example, if we push `Limit` through `Filter`, `Aggregate` and `Distinct`, the results will be wrong. For example, `df.aggregate().limit(1)` and `df.limit(1).aggregate()`will generate the different results.

This is true iff we can push `Limit` through some operators. So far, we did not find any eligible operators except ` outer/left-outer/right-outer Join` and `Union`. Thus, let me revert them back. Thanks!